prov/mrail/src/mrail_cq.c

```diff
@@ -51,6 +82,143 @@ int mrail_cq_write_recv_comp(struct mrail_ep *mrail_ep, struct mrail_hdr *hdr,
 			   NULL, comp->data, hdr->tag);
 }
 
+static int mrail_cq_write_rndv_recv_comp(struct mrail_ep *mrail_ep,
+					 struct mrail_recv *recv)
+{
+	FI_DBG(&mrail_prov, FI_LOG_CQ, "finish rndv recv: length: %zu "
+	       "tag: 0x%" PRIx64 "\n", recv->rndv.len, recv->rndv.tag);
+	ofi_ep_rx_cntr_inc(&mrail_ep->util_ep);
+	if (!(recv->flags & FI_COMPLETION))
+		return 0;
+	return ofi_cq_write(mrail_ep->util_ep.rx_cq, recv->context,
+			   recv->comp_flags | recv->rndv.flags,
+			   recv->rndv.len, NULL, recv->rndv.data,
+			   recv->rndv.tag);
+}
+
+static void mrail_finish_rndv_recv(struct mrail_req *req,
+				   struct fi_cq_tagged_entry *comp)
+{
+	int ret;
+	struct mrail_recv *recv = req->comp.op_context;
+
+	ret = mrail_cq_write_rndv_recv_comp(req->mrail_ep, recv);
+	if (ret) {
+		FI_WARN(&mrail_prov, FI_LOG_CQ,
+			"Cannot write to recv cq\n");
+		assert(0);
+	}
+
+	do {
+		ret = mrail_send_rndv_ack(req->mrail_ep, recv->addr,
+					  (void *)recv->rndv.context);
+		if (ret && ret != -FI_EAGAIN) {
+			FI_WARN(&mrail_prov, FI_LOG_CQ,
+				"Cannot send rndv ack: %s\n", fi_strerror(-ret));
+			assert(0);
+		}
+	} while (ret == -FI_EAGAIN);
```

Comment:
What keeps this from spinning forever?  Does mrail_send_rndv_ack ensure progress?

Reply:
Nothing. I'll see what can be done here.